interceptors_server.go

```diff
@@ -0,0 +1,700 @@
+package gocsi
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"regexp"
+	"sync/atomic"
+
+	"github.com/codedellemc/gocsi/csi"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// ChainUnaryServer chains one or more unary, server interceptors
+// together into a left-to-right series that can be provided to a
+// new gRPC server.
+func ChainUnaryServer(
+	i ...grpc.UnaryServerInterceptor) grpc.UnaryServerInterceptor {
+
+	switch len(i) {
+	case 0:
+		return func(
+			ctx context.Context,
+			req interface{},
+			_ *grpc.UnaryServerInfo,
+			handler grpc.UnaryHandler) (interface{}, error) {
+			return handler(ctx, req)
+		}
+	case 1:
+		return i[0]
+	}
+
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler) (interface{}, error) {
+
+		bc := func(
+			cur grpc.UnaryServerInterceptor,
+			nxt grpc.UnaryHandler) grpc.UnaryHandler {
+			return func(
+				curCtx context.Context,
+				curReq interface{}) (interface{}, error) {
+				return cur(curCtx, curReq, info, nxt)
+			}
+		}
+		c := handler
+		for j := len(i) - 1; j >= 0; j-- {
+			c = bc(i[j], c)
+		}
+		return c(ctx, req)
+	}
+}
+
+// NewServerRequestVersionValidator initializes a new unary server
+// interceptor that validates request versions against the list of
+// supported versions.
+func NewServerRequestVersionValidator(
+	supported []*csi.Version) grpc.UnaryServerInterceptor {
+
+	return (&serverReqVersionValidator{
+		supported: supported,
+	}).handle
+}
+
+type serverReqVersionValidator struct {
+	supported []*csi.Version
+}
+
+func (v *serverReqVersionValidator) handle(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler) (interface{}, error) {
+
+	// Skip version validation if no supported versions are provided.
+	if len(v.supported) == 0 {
+		return handler(ctx, req)
+	}
+
+	treq, ok := req.(hasGetVersion)
+	if !ok {
+		return handler(ctx, req)
+	}
+
+	rv := treq.GetVersion()
+
+	for _, sv := range v.supported {
+		if CompareVersions(rv, sv) == 0 {
+			return handler(ctx, req)
+		}
+	}
+
+	msg := fmt.Sprintf(
+		"unsupported request version: %s", SprintfVersion(rv))
+
+	switch req.(type) {
+	case *csi.ControllerGetCapabilitiesRequest:
+		// UNSUPPORTED_REQUEST_VERSION
+		return ErrControllerGetCapabilities(2, msg), nil
```

Comment:
Hi @codenrhoden,

I use the integers to avoid the gross length of the constants. That's why I also, in the comment above, call out the exact error. I was on the fence about this, but I can switch to the constants. You're not wrong about the possibility of the values changing. However, once the spec is tagged, any changes should be backwards compatible and simply introduce new values.